packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts

```diff
@@ -68,60 +70,170 @@ export enum Protocol {
+  }
+
+  /**
+   * Construct a HTTP2 health check
+   */
+  public static http2(options: HttpHealthCheckOptions = {}): HealthCheck {
+    return new HealthCheckImpl(Protocol.HTTP2, options.healthyThreshold, options.unhealthyThreshold, options.interval, options.timeout, options.path);
+  }
+
+  /**
+   * Construct a GRPC health check
+   */
+  public static grpc(options: GrpcHealthCheckOptions = {}): HealthCheck {
+    return new HealthCheckImpl(Protocol.GRPC, options.healthyThreshold, options.unhealthyThreshold, options.interval, options.timeout);
+  }
+
+  /**
+   * Construct a TCP health check
+   */
+  public static tcp(options: TcpHealthCheckOptions = {}): HealthCheck {
+    return new HealthCheckImpl(Protocol.TCP, options.healthyThreshold, options.unhealthyThreshold, options.interval, options.timeout);
+  }
+
+  /**
+   * Called when the AccessLog type is initialized. Can be used to enforce
+   * mutual exclusivity with future properties
+   */
+  public abstract bind(scope: Construct): HealthCheckConfig;
+}
+
+class HealthCheckImpl extends HealthCheck {
+  constructor(
+    private readonly protocol: Protocol,
+    private readonly healthyThreshold: number = 2,
+    private readonly unhealthyThreshold: number = 2,
+    private readonly interval: cdk.Duration = cdk.Duration.seconds(5),
+    private readonly timeout: cdk.Duration = cdk.Duration.seconds(2),
+    private readonly path?: string) {
+    super();
+    if (healthyThreshold < 2 || healthyThreshold > 10) {
+      throw new Error('healthyThreshold must be between 2 and 10');
+    }
+
+    if (unhealthyThreshold < 2 || unhealthyThreshold > 10) {
+      throw new Error('unhealthyThreshold must be between 2 and 10');
+    }
+
+    if (interval.toMilliseconds() < 5000 || interval.toMilliseconds() > 300_000) {
+      throw new Error('interval must be between 5 seconds and 300 seconds');
+    }
+
+    if (timeout.toMilliseconds() < 2000 || timeout.toMilliseconds() > 60_000) {
+      throw new Error('timeout must be between 2 seconds and 60 seconds');
+    }
+
+    // Default to / for HTTP Health Checks
+    if (path === undefined && (protocol === Protocol.HTTP || protocol === Protocol.HTTP2)) {
+      this.path = '/';
+    }
+  }
+
+  public bind(_scope: Construct): HealthCheckConfig {
+    return {
+      virtualNodeHealthCheck: {
+        protocol: this.protocol,
+        healthyThreshold: this.healthyThreshold,
+        unhealthyThreshold: this.unhealthyThreshold,
+        intervalMillis: this.interval.toMilliseconds(),
+        timeoutMillis: this.timeout.toMilliseconds(),
+        path: this.path,
+      },
+      virtualGatewayHealthCheck: {
+        protocol: this.protocol,
+        healthyThreshold: this.healthyThreshold,
+        unhealthyThreshold: this.unhealthyThreshold,
+        intervalMillis: this.interval.toMilliseconds(),
+        timeoutMillis: this.timeout.toMilliseconds(),
+        path: this.path,
+      },
+    };
```

Comment:
So, this results in dropping the `port` property from the `healthCheck` property of the currently generated CloudFormation templates. Meaning, customers who update their CDK version will see a difference from their previous template when they run `cdk diff`, even though there's no functional change here.

Can I suggest a simple trick that will make this difference go away, without exposing `port` as an input property?

Let's tweak the `HealthCheck` API slightly:

```ts
export interface HealthCheckBindOptions {
  readonly defaultPort?: number;
}

export abstract class HealthCheck
  // ...

  public abstract bind(scope: Construct, options: HealthCheckBindOptions): HealthCheckConfig;
}
```

And then, in `HealthCheckImpl`:

```ts
class HealthCheckImpl extends HealthCheck {
  // ...

  public bind(_scope: Construct, options: HealthCheckBindOptions): HealthCheckConfig {
    return {
      virtualNodeHealthCheck: {
        // remaining properties same as in the PR...
        port: options.port,
      },
      virtualGatewayHealthCheck: {
        // remaining properties same as in the PR...
        port: options.port,
      },
    };
```

And then `bind()` will be called as:

```ts
        healthCheck: this.healthCheck?.bind(scope, { defaultPort: this.port }).virtualGatewayHealthCheck,
```

What do you think @dfezzie?

Reply:
I like that. Makes sense! I'll make the update